File: kalitech-engine/src/main/resources/kalitech/engine/modules/Terrain/helpers/TerrainInstance.js
// FILE: resources/kalitech/builtin/terrain/TerrainInstance.js
"use strict";

class TerrainInstance {
    constructor(api, surface) {
        if (!api) throw new Error("[TERR] api is required");
        if (!surface) throw new Error("[TERR] surface handle is required");

        this._api = api;
        this.surface = surface;

        Object.freeze(this);
    }

    // ------------------------------------------------------------
    // Query
    // ------------------------------------------------------------

    heightAt(x, z, world = true) {
        return this._api.heightAt(this.surface, x, z, world);
    }

    normalAt(x, z, world = true) {
        return this._api.normalAt(this.surface, x, z, world);
    }

    // ------------------------------------------------------------
    // Edit
    // ------------------------------------------------------------

    setHeight(x, z, h, world = true) {
        return this._api.setHeight(this.surface, x, z, h, world);
    }

    adjustHeight(x, z, delta, world = true) {
        return this._api.adjustHeight(this.surface, x, z, delta, world);
    }

    setHeightmap(heights, size, rebuild = true) {
        return this._api.setHeightmap(this.surface, heights, size, rebuild);
    }

    heightmap() {
        return this._api.heightmap(this.surface);
    }

    rebuild() {
        return this._api.rebuild(this.surface);
    }

    // ------------------------------------------------------------
    // Visual / misc
    // ------------------------------------------------------------

    material(mat) {
        return this._api.material(this.surface, mat);
    }

    uv(cfg) {
        return this._api.uv(this.surface, cfg);
    }

    lod(cfg) {
        return this._api.lod(this.surface, cfg);
    }

    scale(xz, cfg) {
        return this._api.scale(this.surface, xz, cfg);
    }

    attach(entityId) {
        return this._api.attach(this.surface, entityId);
    }

    detach() {
        return this._api.detach(this.surface);
    }
}

module.exports = {TerrainInstance};

File: kalitech-engine/src/main/resources/kalitech/engine/modules/Terrain/helpers/TerrainInstance.lua
local M = {}
local luaRuntime = require("@builtin/lua_runtime")
local LuaClass = luaRuntime.LuaClass
local Error = luaRuntime.Error
local LuaConstruct = luaRuntime.LuaConstruct
TerrainInstance = LuaClass()
TerrainInstance.name = "TerrainInstance"
function TerrainInstance.prototype.lua_constructor(self, api, surface)
    if not api then
        error(
            LuaConstruct(Error, "[TERR] api is required"),
            0
        )
    end
    if not surface then
        error(
            LuaConstruct(Error, "[TERR] surface handle is required"),
            0
        )
    end
    self._api = api
    self.surface = surface
    KObject:freeze(self)
end
function TerrainInstance.prototype.heightAt(self, x, z, world)
    if world == nil then
        world = true
    end
    return self._api:heightAt(self.surface, x, z, world)
end
function TerrainInstance.prototype.normalAt(self, x, z, world)
    if world == nil then
        world = true
    end
    return self._api:normalAt(self.surface, x, z, world)
end
function TerrainInstance.prototype.setHeight(self, x, z, h, world)
    if world == nil then
        world = true
    end
    return self._api:setHeight(
        self.surface,
        x,
        z,
        h,
        world
    )
end
function TerrainInstance.prototype.adjustHeight(self, x, z, delta, world)
    if world == nil then
        world = true
    end
    return self._api:adjustHeight(
        self.surface,
        x,
        z,
        delta,
        world
    )
end
function TerrainInstance.prototype.setHeightmap(self, heights, size, rebuild)
    if rebuild == nil then
        rebuild = true
    end
    return self._api:setHeightmap(self.surface, heights, size, rebuild)
end
function TerrainInstance.prototype.heightmap(self)
    return self._api:heightmap(self.surface)
end
function TerrainInstance.prototype.rebuild(self)
    return self._api:rebuild(self.surface)
end
function TerrainInstance.prototype.material(self, mat)
    return self._api:material(self.surface, mat)
end
function TerrainInstance.prototype.uv(self, cfg)
    return self._api:uv(self.surface, cfg)
end
function TerrainInstance.prototype.lod(self, cfg)
    return self._api:lod(self.surface, cfg)
end
function TerrainInstance.prototype.scale(self, xz, cfg)
    return self._api:scale(self.surface, xz, cfg)
end
function TerrainInstance.prototype.attachEntity(self, entityUuid)
    return self._api:attachEntity(self.surface, entityUuid)
end
function TerrainInstance.prototype.attach(self, entityUuid)
    return self:attachEntity(entityUuid)
end
function TerrainInstance.prototype.detach(self)
    return self._api:detach(self.surface)
end
M = {TerrainInstance = TerrainInstance}

return M
